cluster-autoscaler/cloudprovider/gce/templates.go

```diff
@@ -352,19 +372,55 @@ func extractTaintsFromKubeEnv(kubeEnv string) ([]apiv1.Taint, error) {
 }
 
 func extractKubeReservedFromKubeEnv(kubeEnv string) (string, error) {
-	kubeletArgs, err := extractFromKubeEnv(kubeEnv, "KUBELET_TEST_ARGS")
+	// In v1.10+, kube-reserved is only exposed for the autoscaler via AUTOSCALER_ENV_VARS
+	// see kubernetes/kubernetes#61119. We try AUTOSCALER_ENV_VARS first, then
+	// fall back to the old way.
+	kubeReserved, err := extractAutoscalerVarFromKubeEnv(kubeEnv, "kube_reserved")
 	if err != nil {
-		return "", err
+		glog.Infof("kube_reserved not found via AUTOSCALER_ENV_VARS, will try kube-reserved in KUBELET_TEST_ARGS")
+		kubeletArgs, err := extractFromKubeEnv(kubeEnv, "KUBELET_TEST_ARGS")
+		if err != nil {
+			return "", err
+		}
+		resourcesRegexp := regexp.MustCompile(`--kube-reserved=([^ ]+)`)
+
+		for _, value := range kubeletArgs {
+			matches := resourcesRegexp.FindStringSubmatch(value)
+			if len(matches) > 1 {
+				return matches[1], nil
+			}
+		}
+		return "", fmt.Errorf("kube-reserved not in kubelet args in kube-env: %q", strings.Join(kubeletArgs, " "))
 	}
-	resourcesRegexp := regexp.MustCompile(`--kube-reserved=([^ ]+)`)
+	return kubeReserved[0], nil
+}
 
-	for _, value := range kubeletArgs {
-		matches := resourcesRegexp.FindStringSubmatch(value)
-		if len(matches) > 1 {
-			return matches[1], nil
+func extractAutoscalerVarFromKubeEnv(kubeEnv, name string) ([]string, error) {
+	const autoscalerVars = "AUTOSCALER_ENV_VARS"
+	autoscalerVals, err := extractFromKubeEnv(kubeEnv, autoscalerVars)
+	if err != nil {
+		return nil, err
+	}
+	result := []string{}
+	for _, val := range autoscalerVals {
+		for _, v := range strings.Split(val, ";") {
```

Comment:
Can labels or taints contain a semi-colon? If so, we can't use it as the separator char.

Reply:
I don't think so:
labels: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
>Labels are key/value pairs. Valid label keys have two segments: an optional prefix and name, separated by a slash (/). The name segment is required and must be 63 characters or less, beginning and ending with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between. The prefix is optional. If specified, the prefix must be a DNS subdomain: a series of DNS labels separated by dots (.), not longer than 253 characters in total, followed by a slash (/). If the prefix is omitted, the label Key is presumed to be private to the user. Automated system components (e.g. kube-scheduler, kube-controller-manager, kube-apiserver, kubectl, or other third-party automation) which add labels to end-user objects must specify a prefix. The kubernetes.io/ prefix is reserved for Kubernetes core components.
>Valid label values must be 63 characters or less and must be empty or begin and end with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between.

taints: https://github.com/kubernetes/kubernetes/blob/ca02c11887c65431af85f809606527be7f08062d/pkg/apis/core/validation/validation.go#L3931

Taints use the same validation as labels for keys and values. 

The effect part of the taint args to Kubelet (after the `:`) is split out from the value part here when Kubelet parses the flag: https://github.com/kubernetes/kubernetes/blob/ca02c11887c65431af85f809606527be7f08062d/pkg/util/taints/taints.go#L46